### PR TITLE
Allow optionally specifying HTTP request method and body

### PR DESCRIPTION
**Pull request by bendbennett**
References: #21 

`terraform-provider-http` uses a hardcoded HTTP **GET** request method.

The changes contained within this PR allow for optionally specifying:
- request method, one of GET, POST, PUT, PATCH, DELETE
- request body  

These changes are outlined in the [HTTP Verbs](https://docs.google.com/document/d/1mdvfNiNKarz_ipuqffqbQrDsQLmW-GhgXPaFo-NgGFc/edit#heading=h.buibml32t1vx) of the [HTTP Utility Provider Enhancements](https://docs.google.com/document/d/1mdvfNiNKarz_ipuqffqbQrDsQLmW-GhgXPaFo-NgGFc/edit#) RFC.

Closes: #85 
Closes: #52 
Closes: #21

### CODE REVIEW

File: internal/datasources/http/data_source_http.go [hunk 1]
```diff
@@ -39,12 +39,26 @@ the server certificate's chain of trust. Data retrieved from servers not under
...
 			"url": {
 				Description: "The URL for the request. Supported schemes are `http` and `https`.",
 				Type:        types.StringType,
 				Required:    true,
 			},
 
+			"method": {
+				Description: "The HTTP Method for the request. " +
+					"Allowed methods are a subset of methods defined in [RFC7231](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3) namely, " +
+					"`GET`, `POST`, `PUT`, `PATCH` and `DELETE`.",
```
Comment (detro) [inline internal/datasources/http/data_source_http.go:57]: Should we also allow `HEAD` here?

Yes, there is nothing to fetch, but one can use it to:
- confirm an endpoint is reachable
- fetch the headers

Comment (detro) [reply internal/datasources/http/data_source_http.go:57]: I was going to say "don't you also need a planmodifier for setting `GET` as default"... then I realised this is not a resource :(

File: internal/datasources/http/data_source_http.go [hunk 2]
```diff
@@ -178,6 +206,18 @@ func (d *dataSource) Read(ctx context.Context, req tfsdk.ReadDataSourceRequest,
 	resp.Diagnostics.Append(diags...)
 }
 
+var allowedMethods = []string{"GET", "POST", "PUT", "PATCH", "DELETE"}
```
Comment (detro) [inline internal/datasources/http/data_source_http.go:209]: As above, should we add `HEAD`?

**Review by detro — APPROVED** [review]
LGTM, but I left a suggestion to allow also `HEAD`, as I see no harm in it.

File: internal/datasources/http/data_source_http.go (hunk 1 above)
Comment (bendbennett) [reply internal/datasources/http/data_source_http.go]: Fair point. Have added `HEAD`.

File: internal/datasources/http/data_source_http.go (hunk 2 above)
Comment (bendbennett) [reply internal/datasources/http/data_source_http.go]: Fair point. Have added `HEAD`.

**Review by detro — APPROVED** [review]

File: internal/datasources/http/data_source_http.go [hunk 3]
```diff
@@ -100,11 +114,25 @@ func (d *dataSource) Read(ctx context.Context, req tfsdk.ReadDataSourceRequest,
...
+		method = "GET"
+	}
+
+	if !isAllowedMethod(method) {
+		resp.Diagnostics.AddError(
+			"Method not allowed",
+			fmt.Sprintf("Method %q not allowed, must be one of: \"%s\".", method, strings.Join(allowedMethods, `", "`)),
+		)
+		return
+	}
```
Comment (bflad) [inline internal/datasources/http/data_source_http.go:131]: Drive-by note: This can use the shiny new terraform-plugin-framework-validators v0.3.0 functionality now 👍 

```go
Validators: []tfsdk.AttributeValidator{
    stringvalidator.OneOf([]string{"GET", "POST", "PUT", "PATCH", "DELETE", "HEAD"}...),
}
```

Comment (bendbennett) [reply internal/datasources/http/data_source_http.go]: Nice! Good point. Have refactored.

**Comment by pascal-hofmann** [comment]
I need this. Is there anything I can help with to get this merged?

File: internal/provider/data_source_http.go [hunk 4]
```diff
@@ -35,12 +36,36 @@ the server certificate's chain of trust. Data retrieved from servers not under
...
+					"`GET`, `POST`, `PUT`, `PATCH`, `DELETE` and `HEAD`.",
+				Type:     types.StringType,
+				Optional: true,
+				Validators: []tfsdk.AttributeValidator{
+					stringvalidator.OneOf([]string{
+						http.MethodGet,
+						http.MethodPost,
+						http.MethodPut,
+						http.MethodPatch,
+						http.MethodDelete,
```
Comment (bflad) [inline internal/provider/data_source_http.go:63]:
```suggestion
```

File: internal/provider/data_source_http_test.go [hunk 5]
```diff
@@ -277,6 +275,145 @@ func TestDataSource_Provisioner(t *testing.T) {
...
+								method = "DELETE" 
+							}`, testHttpMock.server.URL),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.http.http_test", "response_body", ""),
+					resource.TestCheckResourceAttr("data.http.http_test", "status_code", "204"),
+				),
+			},
+		},
+	})
+}
```
Comment (bflad) [inline internal/provider/data_source_http_test.go:368]:
```suggestion
```

File: internal/provider/data_source_http_test.go [hunk 6]
```diff
@@ -277,6 +275,145 @@ func TestDataSource_Provisioner(t *testing.T) {
...
+		ProtoV5ProviderFactories: protoV5ProviderFactories(),
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(`
+							data "http" "http_test" {
+ 								url = "%s/deleted"
+								method = "OPTIONS" 
+							}`, testHttpMock.server.URL),
+				// Terraform < 0.15 wraps in a  different location to TF >= 0.15 hence the use of (?:\n| ).
+				ExpectError: regexp.MustCompile(`.*Value must be one of: \["\\"GET\\"" "\\"POST\\"" "\\"PUT\\"" "\\"PATCH\\""(?:\n| )"\\"DELETE\\""(?:\n| )"\\"HEAD\\""`),
```
Comment (bflad) [inline internal/provider/data_source_http_test.go:411]:
```suggestion
				ExpectError: regexp.MustCompile(`.*Value must be one of: \["\\"GET\\"" "\\"POST\\"" "\\"HEAD\\""),
```

File: internal/provider/data_source_http_test.go [hunk 7]
```diff
@@ -312,6 +449,29 @@ func setUpMockHttpServer() *TestHttpMock {
...
+				}
+			case "/modified":
+				if r.Method == "PATCH" {
+					w.WriteHeader(http.StatusOK)
+					_, _ = w.Write([]byte("modified"))
+				}
+			case "/deleted":
+				if r.Method == "DELETE" {
+					w.WriteHeader(http.StatusNoContent)
+				}
```
Comment (bflad) [inline internal/provider/data_source_http_test.go:470]:
```suggestion
```

**Review by bflad — COMMENTED** [review]
To prevent any suggestions that this data source should be used for URLs with side-effects, which goes against the general data source concept in Terraform, I think we should pare down the supported methods until we have a better understanding of use cases. I'm guessing those would be best served as a managed resource or in their own provider which understands the semantics of the HTTP API, the necessary data to store for properly refreshing remote state, etc. We can always revisit the additional methods afterwards. 👍 

Otherwise I think this is looking really good.

**Review by bflad — APPROVED** [review]
Looks good to me 🚀
